### PR TITLE
avoid expensive ref counting for CompareInfo on Linux

### PR DESCRIPTION
**Pull request by adamsitnik**
### The problem

In my previous two PRs I have improved the performance of the string operations on Linux but our Linux implementation is still slower than Windows. For example for [this](https://github.com/dotnet/corefx/issues/37691#issuecomment-501166243) benchmark we get 90k RPS vs 60k RPS.

The captured trace file shows that the two hottest methods are:

```log
System.Private.CoreLib!System.Runtime.InteropServices.SafeHandle::InternalRelease(bool)[OptimizedTier1] 
System.Private.CoreLib!System.Runtime.InteropServices.SafeHandle::DangerousAddRef(bool&)[OptimizedTier1]
```

obraz

Those two methods are responsible for ref counting.

The question is: do we need ref counting for this particular handle?

In my opinion, not. 

### SortHandle

What is `SortHandle`: it's a wrapper for an array of pointers to ICU UCollators that are required for culture-specific string operations on Linux in System.Globalization.Native.so.

https://github.com/dotnet/coreclr/blob/c3ac0c460382519c0344f3e87626846ab6fd96a1/src/corefx/System.Globalization.Native/pal_collation.c#L43-L46

It's manged representation was so far a type derived from `SafeHandle`:

https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs#L59-L77

### The Life cycle of `SortHandle`

When new `CompareInfo` is created, it's ctor calls the InitSort method:

https://github.com/dotnet/coreclr/blob/c3ac0c460382519c0344f3e87626846ab6fd96a1/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs#L63

The `InitSort` method calls `Interop.Globalization.GetSortHandle`. If it fails, the handle is `Disposed`

https://github.com/dotnet/coreclr/blob/c3ac0c460382519c0344f3e87626846ab6fd96a1/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L33-L36

The `GetSortHandle` allocates memory for the handle and calls `ucol_open` that creates first `UCollator` and writes it to `collatorsPerOption[0]`

https://github.com/dotnet/coreclr/blob/c3ac0c460382519c0344f3e87626846ab6fd96a1/src/corefx/System.Globalization.Native/pal_collation.c#L358

The cleanup relies on finalizer, the dispose method is not called in explicit way.

The type is internal and not exposed outside System.Private.CoreLib.

### CriticalHandle

The `CriticalHandle` does not perform ref counting on marshaling.
 
With the proposed change, I am able to get from **60k RPS to 110k RPS**

```log
RequestsPerSecond:           110 245
Max CPU (%):                 94
WorkingSet (MB):             177
Avg. Latency (ms):           2,65
Startup (ms):                270
First Request (ms):          110,62
Latency (ms):                0,91
Total Requests:              562 104
Duration: (ms)               5 100
Socket Errors:               0
Bad Responses:               0
SDK:                         3.0.100-preview7-012367
Runtime:                     3.0.0-preview7-27812-01
ASP.NET Core:                3.0.0-preview7.19311.5
```

Before:

obraz

After:

obraz

Please keep in mind that I am not a security expert and my knowledge is limited. If this is a bad idea, please just explain why.

### CODE REVIEW

**Comment by stephentoub** [comment]
@jkoritzinsky, @AaronRobinsonMSFT, does marshaling a CriticalHandle guarantee the object will remain referenced until the P/Invoke returns, akin to having a GC.KeepAlive on the object after the call?  I assume yes, but I want to confirm that the object couldn't be deemed collectible and then finalized while its wrapped handle is still in use by native code.

**Comment by stephentoub** [comment]
Nevermind, answered my own question ("yes"):
https://github.com/dotnet/coreclr/blob/b271aff1fa54c1385143f3b45c1bf3af01c901cd/src/vm/ilmarshalers.cpp#L3107-L3110

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ namespace System.Globalization
     public partial class CompareInfo
     {
         [NonSerialized]
-        private Interop.Globalization.SafeSortHandle _sortHandle = null!; // initialized in helper called by ctors
+        private Interop.Globalization.CriticalSortHandle _sortHandle = null!; // initialized in helper called by ctors
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: Please add a comment that this must not be explicitly Dispose'd.  If it ever needs to be, this will need to change back to being a SafeHandle; otherwise, if it could end up being Dispose'd concurrently with P/Invokes, and we could end up in bad situations.

File: src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs [hunk 2]
```diff
@@ -12,19 +12,20 @@ internal static partial class Interop
     internal static partial class Globalization
     {
         [DllImport(Libraries.GlobalizationNative, CharSet = CharSet.Unicode, EntryPoint = "GlobalizationNative_GetSortHandle")]
-        internal static extern unsafe ResultCode GetSortHandle(byte[] localeName, out SafeSortHandle sortHandle);
+        internal static extern unsafe ResultCode GetSortHandle(byte[] localeName, out CriticalSortHandle sortHandle);
 
+        [SuppressUnmanagedCodeSecurity]
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]: Why is this needed?

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: (I assume we don't explicitly Dispose it because we expect that the vast majority of CompareInfos will be cached / singletons?)

**Comment by jkotas** [comment]
> prevent the CriticalHandle from being finalized during the call-out to native 

This comment is not quite correct. The KeepAlive prevents the object from becoming eligible for finalization, but it does not prevent the handle from being finalized during the call. It is not hard to write a test that resurrects the object in the finalizer and that will fail gracefully with "SafeHandle finalized" exception before this change, but that will crash in unmanaged code after this change.

I know we have been playing it fast and lose with security guarantees like these in name of performance, but it always makes me worried.

**Comment by stephentoub** [comment]
> It is not hard to write a test that resurrects the object in the finalizer

In general, or in this specific case?  Seems like for this specific case, in order to do such resurrection you would need to have access to the object, which would mean such code would need to be in a finalizer on CompareInfo, or in the derived CriticalHandle's finalizer, right?

File: src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]: That shouldn't be needed at all. In fact, I believe we entirely ignore that attribute in .NET Core.

**Comment by AaronRobinsonMSFT** [comment]
>The KeepAlive prevents the object from becoming eligible for finalization, but it does not prevent the handle from being finalized during the call.

@jkotas I don't quite get this statement. If the object isn't eligible for finalization how can it be finalized? Perhaps I am missing what "call" is referring to.

File: src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs (hunk 2 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]: @stephentoub I took this advice from this CriticalHandle comment:

https://github.com/dotnet/coreclr/blob/9773db1e7b1acb3ec75c9cc0e36bd62dcbacd6d5/src/System.Private.CoreLib/shared/System/Runtime/InteropServices/CriticalHandle.cs#L83-L87

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]: I see, thanks. That comment appears to be stale and isn't relevant to .NET Core.

**Comment by jkotas** [comment]
Here is a test that demonstrates the problem: https://gist.github.com/jkotas/91160457be48f8b75eec7c87c003f4bc

It runs fine before this change. It crashed with segmentation fault with this change.

The problem that this is suffering from is https://en.wikipedia.org/wiki/Object_resurrection .

CriticalHandle handle is not effective for solving problems related to object resurrection issues.

**Comment by jkotas** [comment]
As for the solutions for what to do about this:

Majority of the globalization objects are cached as singletons and never freed. The paths where the globalization objects are freed are rare. In fact, it took me quite a while to figure out that I have to pass in `useUserOverride: false` to disable the singleton cache that is required to reproduce the crash. We can make it these sort handles to be *always* cached in singletons and never freed. No need for any sort of smart handles, just store IntPtr and be done with it. It may lead to slightly higher memory consumption in corner cases, but I am not worried about it. It is how it works on Windows today - we never free the sort handle on Windows.

**Comment by tarekgh** [comment]
regarding Windows sort handles, I don't think these are real handle. Sort handle in Windows is just a number they use to index their internal cached data. so there is no any pressure on the system. but in general I agree with @jkotas, most sorting scenarios will just use limited set of cultures with limited set of sort options. so caching and not freeing would be ok as I believe.

**Comment by stephentoub** [comment]
> Here is a test that demonstrates the problem

I see... it's that the critical handle becomes available for finalization not during the initial call but between that and a subsequent call using the resurrected object, and then it can get finalized while it's in use by that subsequent call.

File: src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs (hunk 2 above)
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]: Would be good to fix the comment.

**Comment by adamsitnik** [comment]
@jkotas big thanks for providing a test that demonstrates the problem. BTW are there any good sides of the possibility of having object resurrection?

I have changed the implementation to use `IntPtr`. 

With the changes introduced in this PR, the scenario described in https://github.com/dotnet/corefx/issues/37691 goes from 60k to 110k RPS. 

Following CompareInfo operations should be noticably faster on Linux for small strings:

- CompareString
- IndexOf
- LastIndexOf
- StartsWith
- EndsWith
- GetHashCode

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Windows.cs [hunk 3]
```diff
@@ -427,9 +427,6 @@ private unsafe bool EndsWith(ReadOnlySpan<char> source, ReadOnlySpan<char> suffi
         }
 
         // PAL ends here
-        [NonSerialized]
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Windows.cs]: Nit: I would keep the `_sortHandle` field duplicated between Windows and Unix.

**Comment by jkotas** [comment]
A loop like this will leak memory with the current version of the change:

```
for (;;) {
   new CultureInfo("cz-cz", useUserOverride: false).CompareInfo.GetHashCode(s_shortString, CompareOptions.IgnoreCase);
}
```

To fix this, introduce a singleton Dictionary that will ensure we create the sort handle for given locale exactly once.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs [hunk 4]
```diff
@@ -39,6 +39,9 @@ public partial class CompareInfo : IDeserializationCallback
             ~(CompareOptions.IgnoreCase | CompareOptions.IgnoreSymbols | CompareOptions.IgnoreNonSpace |
               CompareOptions.IgnoreWidth | CompareOptions.IgnoreKanaType | CompareOptions.StringSort);
 
+        [NonSerialized]
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: This lock is only used in the Unix-specific file. It should not be in the common file.

Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: How can I keep it in the Unix-specific file and ensure that it's going to get initialized before the `Invariant` field?

```log
Unhandled exception. System.TypeInitializationException: The type initializer for 'System.Globalization.CompareInfo' threw an exception. ---> System.ArgumentNullException: Value cannot be null.
   at System.Threading.Monitor.ReliableEnter(Object obj, Boolean& lockTaken)
   at System.Globalization.CompareInfo.InitSort(CultureInfo culture)
   at System.Globalization.CultureInfo.get_CompareInfo()
   at System.Globalization.CompareInfo..cctor()
```

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 5]
```diff
@@ -27,17 +37,38 @@ private void InitSort(CultureInfo culture)
...
+                {
+                    if (s_sortNameToSortHandleCache == null)
+                    {
+                        s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+                    }
 
-                    throw new ExternalException(SR.Arg_ExternalException);
+                    if (!s_sortNameToSortHandleCache.TryGetValue(_sortName, out _sortHandle))
+                    {
+                        s_sortNameToSortHandleCache.Add(_sortName, (_sortHandle = GetSortHandle(_sortName)));
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: It would be nice to release the handle if `s_sortNameToSortHandleCache.Add` fails to add it.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 6]
```diff
@@ -30,17 +37,38 @@ private void InitSort(CultureInfo culture)
...
+        private static IntPtr GetSortHandle(string sortName)
+        {
+            var resultCode = Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(sortName), out IntPtr sortHandle);
+            if (resultCode == Interop.Globalization.ResultCode.Success)
+            {
+                return sortHandle;
+            }
+            else
+            {
+                Interop.Globalization.CloseSortHandle(sortHandle);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: This is not correct. GetSortHandle does not return anything on error, and trying to call CloseSortHandle on NULL here will just crash in the native code.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: Create small inner helper type `private static class SortHandleCache` with just the one `Dictionary<string,IntPtr>` static field.

Once you do that, you also do not need the separate `lock`. You can just lock on the Dictionary.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 6 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: good catch, I have removed it

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 5 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: I did not want to use `TryAdd` because I did not know which exception I should throw on failure so I just added a `catch` block for `Add`.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs (hunk 4 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: thanks!

> Once you do that, you also do not need the separate lock object. You can just lock on the Dictionary.

I was wondering why in other types we always have a dictionary and a separate object for lock. Does it mean that if I can provide a good encapsulation I should not be having a separate object for lock?

**Comment by adamsitnik** [comment]
@jkotas I have pushed all the fixes. PTAL one more time. Once again big thanks for the review!

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 7]
```diff
@@ -944,6 +924,64 @@ private SortVersion GetSortVersion()
...
+                    case Interop.Globalization.ResultCode.Success:
+                        return sortHandle;
+                    case Interop.Globalization.ResultCode.OutOfMemory:
+                        throw new OutOfMemoryException();
+                    default:
+                        throw new ExternalException(SR.Arg_ExternalException);
+                }
+            }
+
+            private static byte[] GetNullTerminatedUtf8String(string s)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: This method is not needed. You can just depend on the default interop marshaling that will convert the string to UTF8.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 8]
```diff
@@ -944,6 +924,64 @@ private SortVersion GetSortVersion()
...
+
+                            throw;
+                        }
+                    }
+
+                    return result;
+                }
+            }
+
+            private static IntPtr GetSortHandle(string sortName)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: Nit: I do not think that this being in a separate method is an improvement.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs]: It depends on the situation. If you have a convenient object to lock on, there is no reason to create a separate lock object.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 9]
```diff
@@ -944,6 +924,64 @@ private SortVersion GetSortVersion()
...
+            {
+                switch(Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(sortName), out IntPtr sortHandle))
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]:
```suggestion
                switch (Interop.Globalization.GetSortHandle(GetNullTerminatedUtf8String(sortName), out IntPtr sortHandle))
```

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 7 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: is the default interop marshaling also include the null terminating character?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: yes

**Comment by adamsitnik** [comment]
@jkoritzinsky thank you!  

BTW Having `CharSet.Utf8` would be really nice, it turns out that we even have an issue for that: https://github.com/dotnet/corefx/issues/7804

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 10]
```diff
@@ -944,6 +924,42 @@ private SortVersion GetSortVersion()
...
+            // so caching the sort handles and not freeing them is OK, see https://github.com/dotnet/coreclr/pull/25117 for more
+            private static readonly Dictionary<string, IntPtr> s_sortNameToSortHandleCache = new Dictionary<string, IntPtr>();
+
+            internal static IntPtr GetCachedSortHandle(string sortName)
+            {
+                lock (s_sortNameToSortHandleCache)
+                {
+                    if (!s_sortNameToSortHandleCache.TryGetValue(sortName, out IntPtr result))
+                    {
+                        var resultCode = Interop.Globalization.GetSortHandle(sortName, out result);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]:
```suggestion
                        ResultCode resultCode = Interop.Globalization.GetSortHandle(sortName, out result);
```

The CoreFX coding convention allow 'var' when the type is obvious from the right side.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 11]
```diff
@@ -944,6 +924,42 @@ private SortVersion GetSortVersion()
...
+
+                        if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: Nit: The way this was written originally - check for success first and decide on the exception second - is a tiny bit more efficient. One vs. two conditions on the common path.

File: src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs [hunk 12]
```diff
@@ -11,20 +11,20 @@ internal static partial class Interop
 {
     internal static partial class Globalization
     {
-        [DllImport(Libraries.GlobalizationNative, CharSet = CharSet.Unicode, EntryPoint = "GlobalizationNative_GetSortHandle")]
-        internal static extern unsafe ResultCode GetSortHandle(byte[] localeName, out SafeSortHandle sortHandle);
+        [DllImport(Libraries.GlobalizationNative, CharSet = CharSet.Ansi, EntryPoint = "GlobalizationNative_GetSortHandle")]
+        internal static extern unsafe ResultCode GetSortHandle(string localeName, out IntPtr sortHandle);
 
         [DllImport(Libraries.GlobalizationNative, CharSet = CharSet.Unicode, EntryPoint = "GlobalizationNative_CloseSortHandle")]
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/Interop/Unix/System.Globalization.Native/Interop.Collation.cs]:
```suggestion
        [DllImport(Libraries.GlobalizationNative, EntryPoint = "GlobalizationNative_CloseSortHandle")]
```

It looks odd for Open to be Ansi and Close to be Unicode. Close does not need Chasey at all. It does not deal with strings.

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 10 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: But in this particular case we have to use full type name:

```cs
Interop.Globalization.ResultCode resultCode = Interop.Globalization.GetSortHandle(sortName, out result);
```

imho a `var` combined with good variable name is better:

```cs
var resultCode = Interop.Globalization.GetSortHandle(sortName, out result);
```

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 11 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: In general, I agree that we should avoid having any additional branches on the hot path.

However, in this particular case, the method will be called once or twice per app lifetime (depending on the number of cultures and caching strategy `useUserOverride`) so having two branches is not an issue.

Adding one more `if` would increase the if nesting which imho would be worse.

```cs
lock (s_sortNameToSortHandleCache)
    if (!s_sortNameToSortHandleCache.TryGetValue(sortName, out IntPtr result))
        if (resultCode != resultCode.Success)
            if (resultCode == Interop.Globalization.ResultCode.OutOfMemory)
```

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 10 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: Please don't use var. This name tells me nothing about what kind of result code it is, for example.

Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: I removed the `var`

File: src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 11 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: I do think that the code would actually be more readable and easier to understand with the extra level of nesting.

Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs]: To me, the mapping between the ResultCode and Exception type belongs to ResultCode.

What do you say for the following:

```cs
// part of CompareInfo.Unix.cs
if (resultCode != Interop.Globalization.ResultCode.Success)
    throw resultCode.ToException();

// part of Interop.ResultCode.cs
internal static Exception ToException(this ResultCode resultCode)
{
    Debug.Assert(resultCode != ResultCode.Success);

    switch(resultCode)
    {
        case ResultCode.OutOfMemory:
            return new OutOfMemoryException();
        default:
            return new ExternalException(SR.Arg_ExternalException);
    }
}
```

In this particular use case, the ResultCode is mapped to Exception only once, but if we want to improve the code further this is my proposal.

File: src/System.Private.CoreLib/shared/System/Runtime/InteropServices/CriticalHandle.cs [hunk 13]
```diff
@@ -99,7 +95,7 @@ private MyCriticalHandleSubclass() : base(IntPtr.Zero)
           get { return handle == IntPtr.Zero; }
       }
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Runtime/InteropServices/CriticalHandle.cs]: Not for this PR, but I guess at some point we can remove all these Reliability Contract things.

**Comment by mjsabby** [comment]
@adamsitnik could you please "squash and merge" your changes in future commits? creating a merge commit makes the history non-linear, and all your commits show up in the main history of the repo:

image

**Comment by adamsitnik** [comment]
@mjsabby sure! I always squash the commits and I wanted to squash this PR as well. This time when I hit the "Confirm squash and merge" button I got an error with a message similar to "this PR cannot be merged right now, please refresh the page and try again". I have hit "Confirm squash and merge" again (without refresh) and it got merged. I didn't know that it did not get squashed until I read your post.

Now my "commits of shame" got into coreclr/master and to make it even worse they are going to be ported by the bot account to the "shared" folder of corefx/corert..
